Ed25519.KeyPair.fromSecretKey is broken

### Zig Version

0.10.0

### Steps to Reproduce and Observed Behavior

Try to compile this code (based on the 'ed25519 key pair creation' test):

```zig
const std = @import("std");
const fmt = std.fmt;
const Ed25519 = std.crypto.sign.Ed25519;

test {
    var seed: [32]u8 = undefined;
    _ = try fmt.hexToBytes(seed[0..], "8052030376d47112be7f73ed7a019293dd12ad910b654455798b4667d73de166");
    const key_pair = try Ed25519.KeyPair.create(seed);
    const secret_key = key_pair.secret_key;
    const key_pair_2 = try Ed25519.KeyPair.fromSecretKey(secret_key);
    var buf: [256]u8 = undefined;
    try std.testing.expectEqualStrings(try std.fmt.bufPrint(&buf, "{s}", .{std.fmt.fmtSliceHexUpper(&key_pair_2.secret_key.toBytes())}), "8052030376D47112BE7F73ED7A019293DD12AD910B654455798B4667D73DE1662D6F7455D97B4A3A10D7293909D1A4F2058CB9A370E43FA8154BB280DB839083");
    try std.testing.expectEqualStrings(try std.fmt.bufPrint(&buf, "{s}", .{std.fmt.fmtSliceHexUpper(&key_pair_2.public_key.toBytes())}), "2D6F7455D97B4A3A10D7293909D1A4F2058CB9A370E43FA8154BB280DB839083");
}
```

It does not compile:
```
$ zig test repro.zig
/usr/lib/zig/lib/std/crypto/25519/ed25519.zig:272:26: error: expected type 'error{IdentityElement}', found 'error{InvalidEncoding}'
            const pk_p = try Curve.fromBytes(secret_key.publicKeyBytes());
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/zig/lib/std/crypto/25519/ed25519.zig:272:26: note: 'error.InvalidEncoding' not a member of destination error set
referenced by:
    test_0: repro.zig:10:43
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```

### Expected Behavior

I expected this code to compile. It seems that the return type of the function is not corret (changing it to `EncodingError!KeyPair` works). However, the body of the function also seems to have some issues. I currently patched the function in my local standard library copy like this:

```zig
/// Create a KeyPair from a secret key.
pub fn fromSecretKey(secret_key: SecretKey) EncodingError!KeyPair {
    const pk_p = try Curve.fromBytes(secret_key.publicKeyBytes());

    // It is critical for EdDSA to use the correct public key.
    // In order to enforce this, a SecretKey implicitly includes a copy of the public key.
    // In Debug mode, we can still afford checking that the public key is correct for extra safety.
    //if (std.builtin.mode == .Debug) {
    //    const recomputed_kp = try create(secret_key[0..seed_length].*);
    //    debug.assert(recomputed_kp.public_key.p.toBytes() == pk_p.toBytes());
    //}
    return KeyPair{
        .public_key = PublicKey{ .bytes = pk_p.toBytes() },
        .secret_key = secret_key,
    };
}
```

This is of course not ideal, but it does the trick for me for now.

--- a/lib/std/crypto/25519/ed25519.zig
+++ b/lib/std/crypto/25519/ed25519.zig
@@ -1,3 +1,4 @@
+const builtin = @import("builtin");
 const std = @import("std");
 const crypto = std.crypto;
 const debug = std.debug;
@@ -268,18 +269,17 @@ pub const Ed25519 = struct {
         }
 
         /// Create a KeyPair from a secret key.
-        pub fn fromSecretKey(secret_key: SecretKey) IdentityElementError!KeyPair {
-            const pk_p = try Curve.fromBytes(secret_key.publicKeyBytes());
-
+        pub fn fromSecretKey(secret_key: SecretKey) (NonCanonicalError || EncodingError || IdentityElementError)!KeyPair {
             // It is critical for EdDSA to use the correct public key.
             // In order to enforce this, a SecretKey implicitly includes a copy of the public key.
             // In Debug mode, we can still afford checking that the public key is correct for extra safety.
-            if (std.builtin.mode == .Debug) {
-                const recomputed_kp = try create(secret_key[0..seed_length].*);
-                debug.assert(recomputed_kp.public_key.p.toBytes() == pk_p.toBytes());
+            if (builtin.mode == .Debug) {
+                const pk_p = try Curve.fromBytes(secret_key.publicKeyBytes());
+                const recomputed_kp = try create(secret_key.seed());
+                debug.assert(mem.eql(u8, &recomputed_kp.public_key.toBytes(), &pk_p.toBytes()));
             }
             return KeyPair{
-                .public_key = PublicKey{ .p = pk_p },
+                .public_key = try PublicKey.fromBytes(secret_key.publicKeyBytes()),
                 .secret_key = secret_key,
             };
         }
@@ -673,4 +673,11 @@ test "ed25519 signatures with streaming" {
     verifier.update("mess");
     verifier.update("age");
     try verifier.verify();
+}
+
+test "ed25519 key pair from secret key" {
+    const kp = try Ed25519.KeyPair.create(null);
+    const kp2 = try Ed25519.KeyPair.fromSecretKey(kp.secret_key);
+    try std.testing.expectEqualSlices(u8, &kp.secret_key.toBytes(), &kp2.secret_key.toBytes());
+    try std.testing.expectEqualSlices(u8, &kp.public_key.toBytes(), &kp2.public_key.toBytes());
 }

